### DISCUSSION TITLE
Allowing eventfd

### DISCUSSION THREAD

**Discussion by nmwael** [gramineproject/gramine#1413] [Q&A]
Hi I noticed this:

https://github.com/gramineproject/gramine/blob/fdd3417b490ea0c287fb967432509b09161eb7f3/CI-Examples/rust/rust-hyper-http-server.manifest.template#L34

In our demo setup we have a similar need, as we are running an embedded Apache Tomcat that cannot run without it enabled. How severe are the security concerns behind this. And whats the timeline for gramine implementing it in a secure way?

**Comment by dimakuv** [comment]
> In our demo setup we have a similar need, as we are running an embedded Apache Tomcat that cannot run without it enabled. How severe are the security concerns behind this.

This would require some analysis of how exactly Apache Tomcat uses eventfd. This means reviewing the source code of Apache Tomcat (or its dependency that uses eventfd), and understanding the flows of eventfd. In general, eventfd is just a signaling mechanism between two processes, or between a process and the Linux kernel. If this signaling is somehow verified by the application itself, or if spurious eventfd wake-ups (on select/poll/epoll) can be tolerated by the app, then I would say that there are no security concerns in using `sys.insecure__allow_eventfd`.

> And whats the timeline for gramine implementing it in a secure way?

Eventfd is used in two scenarios:
1. Process to process signaling. This is kinda similar to a pipe between two processes.
2. Kernel to process signaling.

Gramine plans to implement a secure version of eventfd for scenario 1. Gramine **cannot** implement a secure version of eventfd for scenario 2, as the kernel (host Linux) is untrusted in the SGX model, so it's plain impossible.

As for the timeline -- eventfd was always a low priority for us... Check [our Roadmap](https://github.com/orgs/gramineproject/projects/1/views/1) -- looks like we don't even have eventfd task there...

**Reply by nmwael** [reply]
Very interesting answer. I'm going look into this, as I actually have a few different web servers to pick from which have been implemented very different. I will also see if it's something that could be tweaked on with tomcat.

It does seem to be something like you mention that shouldn't cause any security compromises.

**Reply by nmwael** [reply]
After brief look it seems to be a java platform  optimization, which I suspect can be disabled..
